pymc4/variational/approximations.py

```diff
@@ -0,0 +1,212 @@
+"""Implements ADVI approximations."""
+from typing import Optional, Union
+
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from pymc4 import flow
+from pymc4.coroutine_model import Model
+from pymc4.inference.utils import initialize_sampling_state
+
+tfd = tfp.distributions
+tfb = tfp.bijectors
+
+
+class Approximation(object):
+    """Base Approximation class."""
+
+    def __init__(self, model: Optional[Model] = None, random_seed: Optional[int] = None):
+        if not isinstance(model, Model):
+            raise TypeError(
+                "`fit` function only supports `pymc4.Model` objects, but you've passed `{}`".format(
+                    type(model)
+                )
+            )
+
+        self.model = model
+        self._seed = random_seed
+        self.state, self.deterministic_names = initialize_sampling_state(model)
+        if not self.state.all_unobserved_values:
+            raise ValueError(
+                f"Can not calculate a log probability: the model {model.name or ''} has no unobserved values."
+            )
+
+        self.unobserved_keys = self.state.all_unobserved_values.keys()
+        self.target_log_prob = self._build_logfn()
+        self.approx = self._build_posterior()
```

Comment:
Remember, we will need API to convert samples from this distribution to [here](https://github.com/pymc-devs/pymc4/blob/d4936038e32a9dcb8b512d47aed98fddcc6667cc/pymc4/inference/utils.py#L34)

Reply:
The interface now contains a `sample` method which can be used alongside `fit` function after getting the posterior. I have been observing how `trace` from `tfp.mcmc.sample_chain` gets converted to ArviZ InferenceData. Experimenting out with it ([gist](https://gist.github.com/Sayam753/36bf35c482b705545eecb5353a8f8f6a)), ArviZ shows a warning with respect to shapes for Multivariate Distributions, if `az.from_dict` is included in `sample` method. I am still exploring how to resolve that. Any suggestions over this will be helpful.